File: CfxGenerator/ApiTypeBuilder.vb
'' Copyright (c) 2014-2015 Wolfgang Borgsmüller
'' All rights reserved.
'' 
'' Redistribution and use in source and binary forms, with or without 
'' modification, are permitted provided that the following conditions 
'' are met:
'' 
'' 1. Redistributions of source code must retain the above copyright 
''    notice, this list of conditions and the following disclaimer.
'' 
'' 2. Redistributions in binary form must reproduce the above copyright 
''    notice, this list of conditions and the following disclaimer in the 
''    documentation and/or other materials provided with the distribution.
'' 
'' 3. Neither the name of the copyright holder nor the names of its 
''    contributors may be used to endorse or promote products derived 
''    from this software without specific prior written permission.
'' 
'' THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS 
'' "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT 
'' LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS 
'' FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
'' COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, 
'' INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, 
'' BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS 
'' OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND 
'' ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
'' TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE 
'' USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Imports System.IO
Imports System.Runtime.Serialization
Imports System.Runtime.Serialization.Formatters.Binary

Public Class ApiTypeBuilder

    Private apiData As Parser.CefApiData
    Private apiTypes As SortedDictionary(Of String, ApiType)

    Private constStringPtrType As CefStringPtrTypeConst

    Public Function GetDeclarations() As CefApiDeclarations
        apiData = Deserialize()
        If apiData Is Nothing Then
            Dim parser = New Parser.ApiParser
            apiData = parser.Parse()
            Serialize(apiData)
        End If
        Return BuildTypes()
    End Function

    Private Sub Serialize(apiData As Parser.CefApiData)
        Dim formatter = New BinaryFormatter
        Dim stream = New FileStream("cef\SerializedApi.data", FileMode.Create, FileAccess.Write, FileShare.None)
        formatter.Serialize(stream, apiData)
        stream.Close()
    End Sub

    Private Function Deserialize() As Parser.CefApiData

        If Not File.Exists("cef\SerializedApi.data") Then Return Nothing
        Dim formatter = New BinaryFormatter
        Dim stream = New FileStream("cef\SerializedApi.data", FileMode.Open, FileAccess.Read, FileShare.Read)
        Try
            Return DirectCast(formatter.Deserialize(stream), Parser.CefApiData)
        Catch ex As Exception
            Return Nothing
        Finally
            stream.Close()
        End Try
    End Function

    Private Function BuildTypes() As CefApiDeclarations

        Dim structs = New List(Of CefStructType)
        Dim enums = New List(Of CefEnumType)
        Dim stringCollectionTypes = New List(Of StringCollectionType)
        Dim stringCollectionFunctions As New List(Of CefExportFunction)
        Dim functions As New List(Of CefExportFunction)


        apiTypes = New SortedDictionary(Of String, ApiType)

        AddType(New ApiType("void"))

        AddType(New CefBaseType)
        AddType(New CefBasePtrType)

        AddType(New CefColorType)

        AddType(New BooleanInteger)
        AddType(New BooleanIntegerOutType)

        AddType(New OpaquePtrType("XDisplay"))
        AddType(New CefPlatformBasePtrType("cef_window_info_t*"))
        AddType(New CefPlatformBasePtrType("cef_main_args_t*"))

        Dim tStr As ApiType = New CefStringType
        AddType(tStr)
        AddType(New CefStringPtrType(tStr.AsCefStringType))
        constStringPtrType = New CefStringPtrTypeConst(tStr.AsCefStringType)

        AddType(New CefStringUserFreeType)

        NumericType.CreateAll(Me)

        Dim bTypes = AssemblyResources.GetLines("BlittableTypes.txt")
        For Each bt In bTypes
            Dim def = bt.Split("|"c)
            If def.Length = 2 AndAlso def(0).Length > 0 AndAlso def(1).Length > 0 Then
                AddType(New BlittableType(def(0), def(1)))
            End If
        Next

        AddType(New VoidPtrPtrType())

        For Each ed In apiData.CefEnums
            Dim t = New CefEnumType(ed)
            AddType(t)
            enums.Add(t)
        Next

        For Each sd In apiData.CefStructs
            If Not apiTypes.Keys.Contains(sd.Name) Then
                Dim structName = sd.Name.Substring(0, sd.Name.Length - 2)
                Dim t = New CefStructType(structName, sd.Comments)
                AddType(t)
                structs.Add(t)
            End If
        Next

        For Each sd In apiData.CefStructsWindows
            Dim structName = sd.Name.Substring(0, sd.Name.Length - 2)
            Dim t = New CefPlatformStructType(structName, sd.Comments, CefPlatform.Windows)
            AddType(t)
            structs.Add(t)
        Next

        For Each sd In apiData.CefStructsLinux
            Dim structName = sd.Name.Substring(0, sd.Name.Length - 2)
            Dim t = New CefPlatformStructType(structName, sd.Comments, CefPlatform.Linux)
            AddType(t)
            structs.Add(t)
        Next

        stringCollectionTypes.Add(New CefStringListType)
        stringCollectionTypes.Add(New CefStringMapType)
        stringCollectionTypes.Add(New CefStringMultimapType)

        For Each t In stringCollectionTypes
            AddType(t)
        Next

        
        For Each sd In apiData.CefStructs
            Dim t = apiTypes(sd.Name)
            If t.IsCefStructType Then
                t.AsCefStructType.SetMembers(sd, Me)
            End If
        Next

        For Each sd In apiData.CefStructsWindows
            Dim t = apiTypes(sd.Name.Substring(0, sd.Name.Length - 2) & "_windows")
            t.AsCefStructType.SetMembers(sd, Me)
        Next

        For Each sd In apiData.CefStructsLinux
            Dim t = apiTypes(sd.Name.Substring(0, sd.Name.Length - 2) & "_linux")
            t.AsCefStructType.SetMembers(sd, Me)
        Next

        For Each fd In apiData.CefStringCollectionFunctions
            stringCollectionFunctions.Add(New CefExportFunction(Nothing, fd, Me))
        Next

        For Each fd In apiData.CefFunctions
            Dim f = New CefExportFunction(Nothing, fd, Me)
            functions.Add(f)
        Next

        functions.Sort(New CefExportFunction.Comparer)
        structs.Sort(New ApiType.Comparer)
        enums.Sort(New ApiType.Comparer)

        Dim decl = New CefApiDeclarations(functions.ToArray(), structs.ToArray(), enums.ToArray(), stringCollectionTypes.ToArray(), stringCollectionFunctions.ToArray())
        Return decl

    End Function

    Public Sub AddType(t As ApiType)
        For Each m In t.ParserMatches
            apiTypes.Add(m, t)
        Next
    End Sub

    Public Function GetApiType(td As Parser.TypeData, isConst As Boolean) As ApiType

        Dim t = FindApiType(td)

        If t.IsCefStringPtrType AndAlso isConst Then
            Return constStringPtrType
        End If

        Return t
    End Function

    Private Function FindApiType(td As Parser.TypeData) As ApiType

        Dim t As ApiType = Nothing

        If td.Indirection Is Nothing Then
            If Not apiTypes.TryGetValue(td.Name, t) Then
                Stop
            End If
            Return t
        End If

        If apiTypes.TryGetValue(td.Name & td.Indirection, t) Then
            Return t
        End If

        If Not apiTypes.TryGetValue(td.Name, t) Then
            Stop
        End If

        If t.IsCefStructType Then
            If td.Indirection.Count(Function(c As Char) c.Equals("*"c)) = 1 Then
                t = New CefStructPtrType(t.AsCefStructType, td.Indirection)
            ElseIf td.Indirection = "**" Then
                Dim t0 = New CefStructPtrType(t.AsCefStructType, "*")
                t = New CefStructOutType(t0, td.Indirection)
                't = New CefStructRefType(t0, td.Indirection)
                'Debug.Print("StructRef " & t.ToString())
            ElseIf td.Indirection.Count(Function(c As Char) c.Equals("*"c)) = 2 Then
                Dim t0 = New CefStructPtrType(t.AsCefStructType, "*")
                t = New CefStructPtrPtrType(t0, td.Indirection)
            Else
                Stop
            End If
        ElseIf t.IsBlittableType Then
            t = New BlittablePtrType(t.AsBlittableType, td.Indirection)
        Else
            Stop
        End If

        AddType(t)

        Return t
    End Function

End Class

File: CfxGenerator/ApiTypeBuilder.cs
// Copyright (c) 2014-2017 Wolfgang Borgsmüller
// All rights reserved.
// 
// This software may be modified and distributed under the terms
// of the BSD license. See the License.txt file for details.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Runtime.Serialization.Formatters.Binary;
using System.Text;
using Parser;

public class ApiTypeBuilder {

    private Parser.CefApiNode apiNode;

    private SortedDictionary<string, ApiType> apiTypes;

    private CefStringPtrTypeConst constStringPtrType = new CefStringPtrTypeConst();

    public CefApiDeclarations GetDeclarations() {
        apiNode = Deserialize();
        string hash = Parser.Parser.ParseApiHash();
        if(apiNode == null || !hash.Equals(apiNode.ApiHashUniversal)) {
            apiNode = Parser.Parser.Parse();
            apiNode.ApiHashUniversal = hash;
            Serialize(apiNode);
        }
        return BuildTypes();
    }

    private string serializedApi = System.IO.Path.Combine("cef", "SerializedApi.data");

    private void Serialize(Parser.CefApiNode apiNode) {
        var formatter = new BinaryFormatter();
        var stream = new FileStream(serializedApi, FileMode.Create, FileAccess.Write, FileShare.None);
        formatter.Serialize(stream, apiNode);
        stream.Close();
    }

    private Parser.CefApiNode Deserialize() {
        if(!File.Exists(serializedApi))
            return null;
        var formatter = new BinaryFormatter();
        var stream = new FileStream(serializedApi, FileMode.Open, FileAccess.Read, FileShare.Read);
        try {
            return (Parser.CefApiNode)formatter.Deserialize(stream);
        } catch(Exception) {
            return null;
        } finally {
            stream.Close();
        }
    }

    private void ProcessApi() {
        // process c++ findings (compat with previous parser)

        var classes = new Dictionary<string, CefClassNode>(apiNode.CefClasses.Count);
        var funcs = new Dictionary<string, CefCppFunctionNode>();
        foreach(var f in apiNode.CefCppFunctions) {
            if(f.CefConfig.CApiName == null) {
                funcs.Add(CppStyle2CStyle(f.Name), f);
            } else {
                f.CefConfig.CppApiName = f.Name;
                funcs.Add(f.CefConfig.CApiName, f);
            }
        }
        foreach(var c in apiNode.CefClasses) {
            classes.Add(CppStyle2CStyle(c.Name) + "_t", c);
            foreach(var f in c.Methods) {
                if(f.CefConfig.CApiName == null) {
                    if(f.IsStatic && f.Name.EndsWith(c.Name.Substring(3))) {
                        funcs.Add(CppStyle2CStyle(c.Name) + "_" + CppStyle2CStyle(f.Name.Substring(0, f.Name.Length - (c.Name.Length - 3))), f);
                    } else {
                        funcs.Add(CppStyle2CStyle(c.Name) + "_" + CppStyle2CStyle(f.Name), f);
                    }
                } else {
                    f.CefConfig.CppApiName = f.Name;
                    funcs.Add(CppStyle2CStyle(c.Name) + "_" + f.CefConfig.CApiName, f);
                }
            }
        }

        foreach(var s in apiNode.CefCallbackStructs) {
            if(classes.ContainsKey(s.Name)) {
                var c = classes[s.Name];
                classes.Remove(s.Name);
                s.CefConfig = c.CefConfig;
            } else {
                if(s.CefFunctions.Count > 0)
                    Debugger.Break();
            }
            foreach(var c in s.CefCallbacks) {
                if(funcs.ContainsKey(s.Name.Substring(0, s.Name.Length - 1) + c.Name)) {
                    var cf = funcs[s.Name.Substring(0, s.Name.Length - 1) + c.Name];
                    c.CefConfig = cf.CefConfig;
                    ApplyBoolParameters(c.Signature, cf);
                } else {
                    if(c.Signature != null)
                        Debugger.Break();
                }
            }
        }

        foreach(var f in apiNode.CefFunctions) {
            if(funcs.ContainsKey(f.Name)) {
                var cf = funcs[f.Name];
                ApplyBoolParameters(f.Signature, cf);
            } else {
                //Debugger.Break();
            }
        }

        apiNode.CefCallbackStructs.Sort((x, y) => {
            return y.Name.Length - x.Name.Length;
        });

        int ifunc = 0;
        while(ifunc < apiNode.CefFunctions.Count) {
            var f = apiNode.CefFunctions[ifunc];
            var found = false;
            foreach(var s in apiNode.CefCallbackStructs) {
                if(f.Name.StartsWith(s.Name.Substring(0, s.Name.Length - 1))) {
                    s.CefFunctions.Add(f);
                    found = true;
                    break;
                }
            }
            if(found) {
                apiNode.CefFunctions.RemoveAt(ifunc);
            } else {
                ++ifunc;
            }
        }
    }

    private static void ApplyBoolParameters(SignatureNode signature, CefCppFunctionNode cf) {
        if(cf.IsRetvalBoolean) {
            signature.ReturnType.Name = "bool";
        }
        foreach(var pm in cf.BooleanParameters) {
            var success = false;
            foreach(var pm1 in signature.Parameters) {
                if(pm1.Var == pm) {
                    pm1.ParameterType.Name = "bool";
                    success = true;
                    break;
                }
            }
            if(!success) {
                Debugger.Break();
            }
        }
    }

    private static string CppStyle2CStyle(string symbol) {
        var s = new StringBuilder();
        for(int i = 0; i < symbol.Length; ++i) {
            var c = symbol[i];
            if(char.IsUpper(c)) {
                if(s.Length > 0 && char.IsLower(symbol[i - 1])) s.Append("_");
                s.Append(char.ToLowerInvariant(c));
            } else {
                s.Append(c);
            }
        }
        return s.ToString();
    }

    private CefApiDeclarations BuildTypes() {

        ProcessApi();

        var structs = new List<CefStructType>();
        var enums = new List<CefEnumType>();
        var stringCollectionTypes = new List<StringCollectionType>();
        List<CefExportFunction> stringCollectionFunctions = new List<CefExportFunction>();
        List<CefExportFunction> functions = new List<CefExportFunction>();

        apiTypes = new SortedDictionary<string, ApiType>();

        AddType(new ApiType("void"));

        AddType(new CefBaseRefCountedType());
        AddType(new CefBaseScopedType());
        AddType(new CefBaseRefCountedPtrType());

        AddType(new CefColorType());
        AddType(new CefColorRefType());

        AddType(new BooleanInteger());
        AddType(new BooleanIntegerOutType());

        AddType(new OpaquePtrType("XDisplay"));
        AddType(new CefPlatformBasePtrType("cef_window_info_t*"));
        AddType(new CefPlatformBasePtrType("cef_main_args_t*"));

        AddType(new CefStringType());
        AddType(new CefStringPtrType());
        AddType(new CefStringUserFreeType());

        NumericType.CreateAll(this);

        var bTypes = AssemblyResources.GetLines("BlittableTypes.txt");
        foreach(var bt in bTypes) {
            var def = bt.Split('|');
            if(def.Length == 2 && def[0].Length > 0 && def[1].Length > 0) {
                AddType(new BlittableType(def[0], def[1]));
            }
        }

        AddType(new VoidPtrPtrType());

        foreach(var ed in apiNode.CefEnums) {
            var t = new CefEnumType(ed);
            AddType(t);
            enums.Add(t);
        }

        foreach(var s in apiNode.CefCallbackStructs) {
            if(!apiTypes.Keys.Contains(s.Name)) {
                var structName = s.Name.Substring(0, s.Name.Length - 2);
                var t = new CefStructType(structName, s.CefConfig.Source == "client" ? StructCategory.Client : StructCategory.Library);
                AddType(t);
                structs.Add(t);
            }
        }

        foreach(var s in apiNode.CefValueStructs) {
            if(!apiTypes.Keys.Contains(s.Name)) {
                var structName = s.Name.Substring(0, s.Name.Length - 2);
                var t = new CefStructType(structName, StructCategory.Values);
                AddType(t);
                structs.Add(t);
            }
        }

        foreach(var s in apiNode.CefStructsWindows) {
            var structName = s.Name.Substring(0, s.Name.Length - 2);
            var t = new CefPlatformStructType(structName, StructCategory.Values, CefPlatform.Windows);
            AddType(t);
            structs.Add(t);
        }

        foreach(var s in apiNode.CefStructsLinux) {
            var structName = s.Name.Substring(0, s.Name.Length - 2);
            var t = new CefPlatformStructType(structName, StructCategory.Values, CefPlatform.Linux);
            AddType(t);
            structs.Add(t);
        }

        stringCollectionTypes.Add(new CefStringListType());
        stringCollectionTypes.Add(new CefStringMapType());
        stringCollectionTypes.Add(new CefStringMultimapType());

        foreach(var t in stringCollectionTypes) {
            AddType(t);
        }

        foreach(var sd in apiNode.CefCallbackStructs) {
            var t = apiTypes[sd.Name];
            Debug.Assert(t.IsCefStructType);
            t.AsCefStructType.SetMembers(sd, this);
        }

        foreach(var sd in apiNode.CefValueStructs) {
            var t = apiTypes[sd.Name];
            Debug.Assert(t.IsCefStructType);
            t.AsCefStructType.SetMembers(sd, this);
        }

        foreach(var sd in apiNode.CefStructsWindows) {
            var t = apiTypes[sd.Name.Substring(0, sd.Name.Length - 2) + "_windows"];
            t.AsCefStructType.SetMembers(sd, this);
        }

        foreach(var sd in apiNode.CefStructsLinux) {
            var t = apiTypes[sd.Name.Substring(0, sd.Name.Length - 2) + "_linux"];
            t.AsCefStructType.SetMembers(sd, this);
        }

        foreach(var fd in apiNode.CefStringCollectionFunctions) {
            stringCollectionFunctions.Add(new CefExportFunction(null, fd, this));
        }

        foreach(var fd in apiNode.CefFunctions) {
            var f = new CefExportFunction(null, fd, this);
            functions.Add(f);
        }

        if(apiNode.CefFunctionsWindows.Count > 0)
            System.Diagnostics.Debugger.Break();

        foreach(var fd in apiNode.CefFunctionsLinux) {
            var f = new CefExportFunction(null, fd, this, CefPlatform.Linux);
            functions.Add(f);
        }

        functions.Sort((x, y) => string.Compare(x.Name, y.Name));
        structs.Sort((x, y) => string.Compare(x.Name, y.Name));
        enums.Sort((x, y) => string.Compare(x.Name, y.Name));

        var decl = new CefApiDeclarations(functions.ToArray(), structs.ToArray(), enums.ToArray(), stringCollectionTypes.ToArray(), stringCollectionFunctions.ToArray());
        return decl;
    }

    public void AddType(ApiType t) {
        foreach(var m in t.ParserMatches) {
            apiTypes.Add(m, t);
        }
    }

    public ApiType GetApiType(Parser.TypeNode td, bool isConst) {
        var t = FindApiType(td);

        if(t.IsCefStringPtrType && isConst) {
            return constStringPtrType;
        }

        return t;
    }

    private ApiType FindApiType(Parser.TypeNode td) {
        ApiType t = null;

        if(td.Indirection == null) {
            if(!apiTypes.TryGetValue(td.Name, out t)) {
                System.Diagnostics.Debugger.Break();
            }
            return t;
        }

        if(apiTypes.TryGetValue(td.Name + td.Indirection, out t)) {
            return t;
        }

        if(!apiTypes.TryGetValue(td.Name, out t)) {
            System.Diagnostics.Debugger.Break();
        }

        if(t.IsCefEnumType) {
            if(td.Indirection == "*") {
                t = new CefEnumPtrType(t.AsCefEnumType);
            } else {
                System.Diagnostics.Debugger.Break();
            }
        } else if(t.IsCefStructType) {
            if(td.Indirection.Count((char c) => c.Equals('*')) == 1) {
                t = new CefStructPtrType(t.AsCefStructType, td.Indirection);
            } else if(td.Indirection == "**") {
                var t0 = new CefStructPtrType(t.AsCefStructType, "*");
                t = new CefStructOutType(t0, td.Indirection);
                //t = New CefStructRefType(t0, td.Indirection)
                //Debug.Print("StructRef " & t.ToString())
            } else if(td.Indirection.Count((char c) => c.Equals('*')) == 2) {
                var t0 = new CefStructPtrType(t.AsCefStructType, "*");
                t = new CefStructPtrPtrType(t0, td.Indirection);
            } else {
                System.Diagnostics.Debugger.Break();
            }
        } else if(t.IsBlittableType) {
            t = new BlittablePtrType(t.AsBlittableType, td.Indirection);
        } else {
            System.Diagnostics.Debugger.Break();
        }

        AddType(t);

        return t;
    }
}